UbuntuDrivers/detect.py

```diff
@@ -303,10 +346,46 @@ def _pkg_get_support(apt_cache, pkg):
     return support
 
 
-def _is_runtimepm_supported(pkg, alias):
+def _is_nv_allowing_runtimepm_supported(alias, ver):
+    '''alias: e.g. pci:v000010DEd000024BAsv0000103Csd000089C6bc03sc00i00'''
+    result = re.search('pci:v0000(.*)d0000(.*)sv(.*)', alias)
+    vid = result.group(1)
+    did = result.group(2)
+    if vid != "10DE":
+        return False
+    did = "0x%s" % did
+    path = customized_supported_gpus_json
+    try:
+        with open(path, "r") as stream:
+            try:
+                gpus = list(json.load(stream)['chips'])
+                for gpu in gpus:
+                    if gpu['devid'] == did and 'runtimepm' in gpu['features']:
+                        if gpu['branch'].split('.')[0] != ver:
+                            logging.debug('Candidate version does not match %s != %s' %
+                                          (gpu['branch'].split('.')[0], ver))
+                            return False
+                        logging.info("Found runtimepm supports on %s." % did)
+                        return True
+            except Exception:
+                logging.debug('_is_nv_allowing_runtimepm_supported(): unexpected json detected')
+                pass
+    except Exception:
+        logging.debug('_is_nv_allowing_runtimepm_supported(): unable to read %s' % path)
+        pass
+    return False
+
+
+def _is_runtimepm_supported(apt_cache, pkg, alias):
     '''Check if the package supports runtimepm for the given modalias'''
     try:
-        m = pkg.candidate.record['PmAliases']
```

Comment:
What is the purpose of this change?

Reply:
@tseliot 

From what I read (and test) from the [apt_pkg document](https://apt-team.pages.debian.net/python-apt/library/apt_pkg.html)

the pkg doesn't have the `.candidate` attribute (API changed).
Instead, to get the candidate, we need to use the following code.

In current code from ubuntu-archive, this function will always go to `except`

Reply:
Ok, I wasn't aware of that, thanks.